cytominer_database/tableSchema.py

```diff
@@ -0,0 +1,252 @@
+import click
+import warnings
+import zlib
+import pandas as pd
+import backports.tempfile
+import sqlalchemy.exc
+from sqlalchemy import create_engine
+import pyarrow
+import pyarrow.parquet as pq
+import pyarrow.csv
+import numpy as np
+import collections
+import numpy as np
+import cytominer_database
+import cytominer_database.utils
+import cytominer_database.load
+
+################################################################################
+# Contains "open_writers()"" to generate the dictionary containing the
+# reference data ("writer dictionary"), from which the table schemata will be generated.
+# The reference tables can either be loaded directly from a designated folder,
+# or sampled across all available data, as specified in the config_file.
+#
+# Also contains helper functions "get_grouped_table_paths()" and "get_dict_of_paths()"
+# to generate a list of paths
+# and the sampling function get_reference_paths().
+################################################################################
+
+
+
+def open_writers(source, target, config_file, skip_image_prefix=True):
+    """
+    Determines, loads reference tables and openes them as ParquetWriters.
+    Returns a dictionary referencing the writers.
+    :param source: path to directory containing all parent folders of .csv files
+    :target: output file path
+    :config_file: parsed configuration data (output from cytominer_database.utils.read_config(config_path))
+    :skip_image_prefix: Boolean value specifying if the column headers of
+     the image.csv files should be prefixed with the table name ("Image").
+    """
+    writers_dict = {}  # nested dictionary: dict in dict
+    engine = config_file["ingestion_engine"]["engine"]
+    reference = config_file["schema"]["reference_option"]
+
+    print("engine = ", engine)
+    print("reference = ", reference)
+
+    if engine == "SQLite": # no reference table needed
+        return writers_dict
+
+    if reference == "sample": 
+        # Idea: sample from all tables contained in the subdirectories of source.
+        # Get all possible reference directories for each table kind, as stored in the dictionary
+        directories = sorted(list(cytominer_database.utils.find_directories(source)))
+        grouped_full_paths = get_grouped_table_paths_from_directory_list(directories)
+        # get sample size (as fraction of all available files) from config file
+        ref_fraction = float(config_file["schema"]["ref_fraction"])
+        # build reference dictionary
+        ref_dir = get_reference_paths(ref_fraction, grouped_full_paths)
+
+        # ---------------- print statements ----------------
+        # print("open_writers: reference == 'sample' ")
+        # print("ref_fraction", ref_fraction)
+        # print("grouped_full_paths", grouped_full_paths)
+        # print("ref_dir", ref_dir)
+        # --------------------------------------------------
+
+    else: # elif os.path.isdir(os.path.join(source, reference))
```

Comment:
probably can delete this comment, right?
```suggestion
    else:
```

Reply:
tableSchema.py has been cleaned